### ISSUE TITLE
make the input k in rot90 a list of int to rotate tensors individually in a batch

### ISSUE THREAD

**Issue by Guo-Jian-Wang** [pytorch/pytorch#105062]
### 🚀 The feature, motivation and pitch

I'm using torch.rot90(input, k=1, dims=[0, 1]) to rotate a batch sample. Since the k here should be an int, which means all the samples in a batch will be rotated with the same angle (because they share the same k). What I want to do is to rotate each sample of the batch separately because this will improve the accuracy of my task.

Is it possible to extend rot90 to make the input k a list of int, to make it possible to rotate tensors individually in a batch? 

### Alternatives

A simple alternative to achieve the function described above is to use the for loop: 

import numpy as np
import torch

batch_sample = torch.randn(10, 1, 32, 32, 129)
for i in range(len(batch_sample)):
    k = np.random.choice(4, 1, replace=False)[0]
    batch_sample[i, :] = torch.rot90(batch_sample[i, :], k=k, dims=[1,2])

But it is slow using the for loop. So, I'm wondering if it is possible to extend rot90 to make the input k a list of int, making it faster to rotate tensors individually in a batch.

**Comment by albanD** [comment]
> A simple alternative to achieve the function described above is to use the for loop:

Note that since there are only 4 possible values, it will be more efficient to replace the for-loop by computing all 4 possible results and then combining them with a mask based on the value of k for each sample.

**Comment by Guo-Jian-Wang** [comment]
> Such feature would be cool! In the meantime if you can work with 4D tensors there's a workaround using a grid sampler. Here's how:
> 
> ```python
> import torch
> import torch.nn.functional as F
> 
> def batched_rot90(input, ks):
>     assert len(input) == len(ks), "Number of ks and batch size are not equal"
>     rot_mat = torch.stack([torch.tensor([[torch.cos(k*torch.pi/2), -torch.sin(k*torch.pi/2), 0],
>                                          [torch.sin(k*torch.pi/2), torch.cos(k*torch.pi/2), 0]]) for k in torch.tensor(ks)])
>     grid = F.affine_grid(rot_mat, input.size())
>     rotated_input = F.grid_sample(input, grid)
>     return rotated_input
> ```
> 
> Note that rotation is on last 2 dims

Thanks @GilPinsky, this works for 4D tensors, but my data is 5D tensor with shape of (N, C, D, H, W). To make it works, I expand your function above to be:

`def batch_rot90_3d(obs, ks=None, dims=[2,3]):
    obs_size = obs.size()
    assert obs_size[dims[0]]==obs_size[dims[1]], "The input tensor should have the same length on the dims axis."
    
    if ks is None:
        ks = np.random.choice(4, len(obs), replace=True)
    if np.all(np.sort(dims)==[2,3]):
        # rotate along the X-axis
        rot_mat = torch.stack([torch.tensor([[1, 0, 0, 0],
                                             [0, np.cos(k*np.pi/2), -np.sin(k*np.pi/2), 0],
                                             [0, np.sin(k*np.pi/2),  np.cos(k*np.pi/2), 0]], dtype=torch.float).cuda() for k in ks])
    elif np.all(np.sort(dims)==[2,4]):
        # rotate along the Y-axis
        rot_mat = torch.stack([torch.tensor([[np.cos(k*np.pi/2), 0, np.sin(k*np.pi/2), 0],
                                             [0, 1, 0, 0],
                                             [-np.sin(k*np.pi/2), 0, np.cos(k*np.pi/2), 0]], dtype=torch.float).cuda() for k in ks])
    elif np.all(np.sort(dims)==[3,4]):
        # rotate along the Z-axis
        rot_mat = torch.stack([torch.tensor([[np.cos(k*np.pi/2), -np.sin(k*np.pi/2), 0, 0],
                                             [np.sin(k*np.pi/2), np.cos(k*np.pi/2), 0, 0],
                                             [0, 0, 1, 0]], dtype=torch.float).cuda() for k in ks])
    grid = F.affine_grid(rot_mat, obs_size, align_corners=False)
    rotated_obs = F.grid_sample(obs, grid, align_corners=False)
    return rotated_obs
    `

This function works for 5D tensors, but it is still slower than that when using torch.rot90 with a for loop. Besides, there is an issue when using this function: the rotated tensor should be the same as the original one when k equals 0 or 4, but it is not when using this function. On the contrary, there are no problems when using torch.rot90.

For example,
`s = torch.randn(3,1,32,32,129).cuda()
torch.min(s - batch_rot90_3d(s,ks=[0])) DO NOT equal to 0.
`
This tiny bias will be expanded in the training process.

**Comment by Guo-Jian-Wang** [comment]
> > A simple alternative to achieve the function described above is to use the for loop:
> 
> Note that since there are only 4 possible values, it will be more efficient to replace the for-loop by computing all 4 possible results and then combining them with a mask based on the value of k for each sample.

Thanks @albanD , that is a good idea, but it will use more GPU memory, which is unsuitable for tasks with high-resolution images.
